### PR TITLE
Added SelectMenus

### PR DESCRIPTION
**Pull request by Morazzer**
Created a implementation for the [select menus](https://discord.com/developers/docs/interactions/message-components#select-menus).

Issue for that #796

### CODE REVIEW

File: javacord-core/src/main/java/org/javacord/core/interaction/SelectMenuInteractionImpl.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+     * @param channel  The channel in which the interaction happened. Can be {@code null}.
+     * @param jsonData The json data of the interaction.
+     */
+    public SelectMenuInteractionImpl(DiscordApiImpl api, TextChannel channel, JsonNode jsonData) {
+        super(api, channel, jsonData);
+
+        ObjectNode messageObject = (ObjectNode) jsonData.get("message");
+        ArrayNode componentsObject = (ArrayNode) messageObject.get("components");
+
+        ObjectNode dataObject = (ObjectNode) jsonData.get("data");
```
Comment (KILLEliteMaste) [inline javacord-core/src/main/java/org/javacord/core/interaction/SelectMenuInteractionImpl.java:38]: Same as above

**Review by KILLEliteMaste — CHANGES_REQUESTED** [review]
Thank you for your fist and also quiet big contribution! :)

**Comment by Morazzer** [comment]
image
Is this a problem with my code?

**Comment by Bastian** [comment]
The GitHub Action did not fail, so probably not.
I triggered a rebuild, let's see.

File: javacord-api/src/main/java/org/javacord/api/listener/interaction/SelectMenuChooseListener.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+import org.javacord.api.listener.user.UserAttachableListener;
+
+/**
+ * This listener listens to select menu choices.
+ */
+@FunctionalInterface
+public interface SelectMenuChooseListener extends ServerAttachableListener, UserAttachableListener,
+        TextChannelAttachableListener, GloballyAttachableListener, ObjectAttachableListener {
+
+
```
Comment (KILLEliteMaste) [inline javacord-api/src/main/java/org/javacord/api/listener/interaction/SelectMenuChooseListener.java:17]: Remove line

File: javacord-core/src/main/java/org/javacord/core/entity/message/component/internal/SelectMenuOptionBuilderDelegateImpl.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+import org.javacord.api.entity.message.component.SelectMenuOption;
+import org.javacord.api.entity.message.component.internal.SelectMenuOptionBuilderDelegate;
+import org.javacord.core.entity.message.component.SelectMenuOptionImpl;
+
+public class SelectMenuOptionBuilderDelegateImpl implements SelectMenuOptionBuilderDelegate {
+    private String label;
+    private String value;
+    private String description;
+    private boolean isDefault = false;
+    private Emoji emoji;
```
Comment (KILLEliteMaste) [inline javacord-core/src/main/java/org/javacord/core/entity/message/component/internal/SelectMenuOptionBuilderDelegateImpl.java:13]: Assign `null` to all variables except for isDefault

File: javacord-core/src/main/java/org/javacord/core/entity/message/component/SelectMenuImpl.java [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+        object.put("custom_id", this.customId);
+
+        ArrayNode arrayNode = JsonNodeFactory.instance.arrayNode();
+
+        for (SelectMenuOption option: options) {
+            arrayNode.add(((SelectMenuOptionImpl) option).toJson());
+        }
+
+        object.set("options", arrayNode);
+        object.put("placeholder", this.placeholder);
```
Comment (KILLEliteMaste) [inline javacord-core/src/main/java/org/javacord/core/entity/message/component/SelectMenuImpl.java:121]: Add a `placeholder != null` check

File: javacord-core/src/main/java/org/javacord/core/entity/message/component/SelectMenuOptionImpl.java [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+                emojiObj.put("name", emoji.asUnicodeEmoji().get());
+            }
+
+            object.set("emoji", emojiObj);
+        }
+
+        object.put("label", label);
+        object.put("value", value);
+        object.put("default", isDefault);
+        object.put("description", description);
```
Comment (KILLEliteMaste) [inline javacord-core/src/main/java/org/javacord/core/entity/message/component/SelectMenuOptionImpl.java:113]: Add a `description != null` check

File: javacord-core/src/main/java/org/javacord/core/entity/message/component/SelectMenuOptionImpl.java [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+
+    /**
+     * Gets the select menu option as a {@link ObjectNode}. This is what is sent to Discord.
+     *
+     * @return The select menu option as a ObjectNode.
+     */
+    public ObjectNode toJson() {
+        ObjectNode object = JsonNodeFactory.instance.objectNode();
+
+        if (!(emoji == null)) {
```
Comment (KILLEliteMaste) [inline javacord-core/src/main/java/org/javacord/core/entity/message/component/SelectMenuOptionImpl.java:98]: `emoji != null` is more readable :)

File: javacord-core/src/main/java/org/javacord/core/entity/message/component/internal/SelectMenuOptionBuilderDelegateImpl.java [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+
```
Comment (KILLEliteMaste) [inline javacord-core/src/main/java/org/javacord/core/entity/message/component/internal/SelectMenuOptionBuilderDelegateImpl.java:14]: Remove line

File: javacord-api/src/main/java/org/javacord/api/entity/message/component/SelectMenu.java [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+     * @return The select menu's min values.
+     */
+    int getMinValues();
+
+    /**
+     * Gets the maximum amount of options which can be selected.
+     *
+     * @return The select menu's max values.
+     */
+    int getMaxValues();
```
Comment (KILLEliteMaste) [inline javacord-api/src/main/java/org/javacord/api/entity/message/component/SelectMenu.java:34]: Imo renaming them to `getMinimumValues` and `getMaximumValues` sounds better. Please apply the renaming in all classes

**Review by KILLEliteMaste — CHANGES_REQUESTED** [review]
Just a few more things :)

File: javacord-api/src/main/java/org/javacord/api/entity/message/component/SelectMenu.java (hunk 8 above)
Comment (Morazzer) [reply javacord-api/src/main/java/org/javacord/api/entity/message/component/SelectMenu.java:34]: Also the local variables?

Comment (KILLEliteMaste) [reply javacord-api/src/main/java/org/javacord/api/entity/message/component/SelectMenu.java:34]: Yes

**Review by KILLEliteMaste — APPROVED** [review]

**Comment by KILLEliteMaste** [comment]
Can you please squash the PR into a single commit?

**Comment by Morazzer** [comment]
I hope it's right how i done it

**Comment by KILLEliteMaste** [comment]
Yeah, it's fine. Thank you!

**Pull request merged**